CDAP-4860 Introduce /available endpoint for services, return OK only when discoverable

gokulavasan opened the pull request:
JIRA : https://issues.cask.co/browse/CDAP-4860

Build : http://builds.cask.co/browse/CDAP-DUT4172

gokulavasan committed:
CDAP-4860 Introduce /active endpoint for services to return OK only when discoverable service endpoints are available

Add private constructor to ServiceDiscoverable class

Address PR comments

Rename classes to reflect the new endpoints

Address PR comment - Remove 400 status and return 503 even if the service is stopped

Fix a failing test

Doc changes

Update service.rst

Edits.

anwar6953 commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java:
```diff
@@ -976,6 +983,32 @@ public void getServiceInstances(HttpRequest request, HttpResponder responder,
   }
 
   /**
+   * Return the
```
incomplete javadoc

anwar6953 commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java:
```diff
@@ -976,6 +983,32 @@ public void getServiceInstances(HttpRequest request, HttpResponder responder,
+    ProgramId programId = Ids.namespace(namespaceId).app(appId).service(serviceId);
+    ProgramStatus status = lifecycleService.getProgramStatus(programId);
+    if (status == ProgramStatus.STOPPED) {
+      responder.sendString(HttpResponseStatus.BAD_REQUEST, "Service is stopped. Please start it.");
```
Why should this be a bad request?
Wouldn't it be a valid request to ask simply 'is it active?' against a stopped service?
I imagine the response would be 200 - 'Not Running' or something, right?

poornachandra replied:
Since the client is asking whether a service is active, the right response code should be 503 - stopped, right?

gokulavasan replied:
@anwar6953  @poornachandra  So from the [wiki page](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes) for http response codes, I see the following for 503 : 
`503 Service Unavailable
The server is currently unavailable (because it is overloaded or down for maintenance). Generally, this is a temporary state`. 

In this case, until the user starts the service, the active/available state is not going to change and so I thought it might make sense to return 400 to nudge the user to start the service before hitting this endpoint. That was my reasoning, but I can change it to 503 if that makes more sense.

@anew Any thoughts?

anew replied:
I think it is a 503. It is a valid service request, but the service is down. Matches the description for a 503 pretty precisely.

anew replied:
Also, if you stop and then restart a service, it will be "not running" (that is code will come here) for a short time, preceded and followed by time where it is running but shutting down or starting up. To an external caller, all three states should be the same, I think it would be more confusing if  the client gets 503 back, then retries and now gets a 400, after which it should - strictly speaking - not try again.

gokulavasan replied:
From the 503 explanation, I was mislead by this - `Generally, this is a temporary state`. So I thought if the service is stopped, then it is not a temporary state, until it is started. 

Regarding, restarting, I agree that might be confusing. But since we don't have a restart request in for non system services. So the user has to make two different calls?

I will change it to 503 since it looks like we don't get much benefit out of having that separate status code.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java:
```diff
@@ -976,6 +983,32 @@ public void getServiceInstances(HttpRequest request, HttpResponder responder,
+                               @PathParam("app-id") String appId,
+                               @PathParam("service-id") String serviceId) throws NotFoundException {
+    ProgramId programId = Ids.namespace(namespaceId).app(appId).service(serviceId);
+    ProgramStatus status = lifecycleService.getProgramStatus(programId);
```
should this check whether the program exists? It looks like lifecycleService does not, it only checks existence of the app. Perhaps add it there.

anew replied:
Oh it does. Never mind.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java:
```diff
@@ -976,6 +983,32 @@ public void getServiceInstances(HttpRequest request, HttpResponder responder,
+      EndpointStrategy endpointStrategy = new RandomEndpointStrategy(discoveryServiceClient.discover(serviceName));
+      if (endpointStrategy.pick(300L, TimeUnit.MILLISECONDS) == null) {
+        LOG.trace("Discoverable endpoint {} not found", serviceName);
+        responder.sendString(HttpResponseStatus.SERVICE_UNAVAILABLE, "Service is not active yet. Please try again.");
```
this is not always true. It might be shutting down, or it failed and got restarted. I would say "service is running but not accepting requests" or something like that.

anew commented on cdap-cli/src/main/java/co/cask/cdap/cli/command/GetServiceActiveStatusCommand.java:
```diff
@@ -0,0 +1,75 @@
+
+  @Override
+  public String getPattern() {
+    return String.format("get active service <%s>", ArgumentName.SERVICE);
```
maybe "check service availability"? Perhaps the endpoint should also be "available" instead of "active".

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java:
```diff
@@ -976,6 +983,32 @@ public void getServiceInstances(HttpRequest request, HttpResponder responder,
+   * Return the
+   */
+  @GET
+  @Path("/apps/{app-id}/services/{service-id}/active")
```
or /available ?

gokulavasan replied:
Sure, 'available' is better. I will make that change.

gokulavasan commented:
@anew Addressed comments. Please take a look again and also it would be great to get your thoughts on this discussion : https://github.com/caskdata/cdap/pull/5818#discussion_r63981678

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java:
```diff
@@ -976,6 +983,33 @@ public void getServiceInstances(HttpRequest request, HttpResponder responder,
+      if (endpointStrategy.pick(300L, TimeUnit.MILLISECONDS) == null) {
+        LOG.trace("Discoverable endpoint {} not found", serviceName);
+        responder.sendString(HttpResponseStatus.SERVICE_UNAVAILABLE,
+                             "Service is running but not accepting requests yet.");
```
"yet" may be inaccurate, it may also be shutting down, in which case the correct phrasing would be "not any more". Perhaps just say "at this time".

gokulavasan replied:
will make that change.

anew commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ProgramLifecycleHttpHandlerTest.java:
```diff
@@ -790,13 +791,27 @@ public void testServices() throws Exception {
+
+    activeResponse = getActiveResponse(service2);
+    // Service has not been started, so it should return 400
+    Assert.assertEquals(HttpResponseStatus.BAD_REQUEST.getCode(), activeResponse.getStatusLine().getStatusCode());
```
503

anew commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ProgramLifecycleHttpHandlerTest.java:
```diff
@@ -831,6 +846,10 @@ public void testServices() throws Exception {
+
+    activeResponse = getActiveResponse(service2);
+    // Service has been stopped, so it should return 400
+    Assert.assertEquals(HttpResponseStatus.BAD_REQUEST.getCode(), activeResponse.getStatusLine().getStatusCode());
```
503

anew commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ProgramLifecycleHttpHandlerTest.java:
```diff
@@ -790,13 +791,27 @@ public void testServices() throws Exception {
-                                          APP_WITH_SERVICES_SERVICE_NAME);
+    final Id.Service service2 = Id.Service.from(Id.Namespace.from(TEST_NAMESPACE2), APP_WITH_SERVICES_APP_ID,
+                                                APP_WITH_SERVICES_SERVICE_NAME);
+    HttpResponse activeResponse = getActiveResponse(service1);
```
rename everything that has "active" in its name to "available"? Makes the code easier to read.

gokulavasan replied:
sorry missed this. will rename it

anew commented:
@gokulavasan I made another pass. We are converging.

anew commented:
LGTM

gokulavasan commented:
@anew Thanks for the review! @awholegunch I have made the necessary doc changes in the second commit. Can you please review the same?

awholegunch commented:
Docs LGTM. Please check my changes.

gokulavasan commented:
@awholegunch Looks great. Thanks! I will squash and merge the PR once the build passes.
